pkg/mod/waifu_vault.go

```diff
@@ -1,9 +1,20 @@
 package mod
 
+import "context"
+
 type Waifuvalt interface {
-	UploadFile(options WaifuvaultPutOpts) (*WaifuResponse[string], error)
-	FileInfo(token string) (*WaifuResponse[int], error)
-	FileInfoFormatted(token string) (*WaifuResponse[string], error)
-	DeleteFile(token string) (bool, error)
-	GetFile(options GetFileInfo) ([]byte, error)
+	// UploadFile - Upload a file using a byte array, url or file
+	UploadFile(options WaifuvaultPutOpts, ctx *context.Context) (*WaifuResponse[string], error)
+
+	// FileInfo - Obtain file info such as URL and retention period (as an epoch timestamp)
+	FileInfo(token string, ctx *context.Context) (*WaifuResponse[int], error)
+
+	// FileInfoFormatted - Same as FileInfo, but instead returns the retention period as a human-readable string
+	FileInfoFormatted(token string, ctx *context.Context) (*WaifuResponse[string], error)
+
+	// DeleteFile - Delete a file given a token
+	DeleteFile(token string, ctx *context.Context) (bool, error)
+
+	// GetFile - Download the file given options and return a byte array of said file
+	GetFile(options GetFileInfo, ctx *context.Context) ([]byte, error)
```

Comment:
`ctx` should be always as first argument of methods.
Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:
```go
func DoSomething(ctx context.Context, arg Arg) error {
	// ... use ctx ...
}
```

https://pkg.go.dev/context

Reply:
Also it shouldn't be with pointer, just a `ctx context.Context`.

Apart code you wanna make it optional? Please don't. Always require context.

Reply:
no. i will not make context mandatroy

Reply:
> I am not storing the context in the strict, as you can see very clearly, the context is being passed on each function; but i am not going to make it a requirement.

Then close the PR, thats how golang code should be written.

Reply:
> > I am not storing the context in the strict, as you can see very clearly, the context is being passed on each function; but i am not going to make it a requirement.
> 
> Then close the PR, thats how golang code should be written.

that is how YOU think go should be written. might as well quit my job while i am at it :) if that was the case, then the mtehod `NewRequest` would not exist and only `NewRequestWithContext` would exist.

Reply:
That what i meant, you are using context, or you are not using context, there is no option to pass it as nil for your library.

Try to find some good grade 3rd party lib with pointer context, you cannot.
And thats not how i think, thats how it is described in godocs.

From https://pkg.go.dev/context: 
1. `Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:`
2. `Do not pass a nil Context, even if a function permits it. Pass context.TODO if you are unsure about which Context to use.`
3. `Use context Values only for request-scoped data that transits processes and APIs, not for passing optional parameters to functions.`
4. `The same Context may be passed to functions running in different goroutines; Contexts are safe for simultaneous use by multiple goroutines.`

If you wanna make them optional, them better to not implement them at all, just close PR and thats totally ok.

Reply:
This really does not need a context. the docs say to not pass nil contexts and always keep it as the first argument. but (forgive the irony), context around that is scoped to contextual methods. 

Yeh, this PR is technically wrong, but not because this is defining a context incorrectly. just don't use any contexts. if people REALLY are going to complain about not being able to set delays, timeout channels or cancellation channels, then they can complain. for such a simple thing. don't bother

Reply:
and No, not every API call needs a channel, that is nonsense

Reply:
ok. closing this PR then.